CompuCell3D/core/CompuCell3D/plugins/Chemotaxis/ChemotaxisPlugin.cpp

```diff
@@ -556,6 +558,82 @@ double ChemotaxisPlugin::regularChemotaxis(const Point3D &pt, const CellG *newCe
+	// Equivalent to regularChemotaxis(pt, newCell, oldCell) + regularChemotaxis(pt, oldCell, newCell), but in a single loop over fields
+	// return regularChemotaxis(pt, newCell, oldCell) + regularChemotaxis(pt, oldCell, newCell);
+
+	double energy = 0;
+
+	std::map<std::string,ChemotaxisData>::iterator mitr;
+
+	for(unsigned int i = 0; i < fieldVec.size(); ++i){
+		if(newCell){
+			bool chemotaxisDone = false;
+
+			// first try "locally defined" chemotaxis
+			std::map<std::string,ChemotaxisData> & chemotaxisDataDictRef = *chemotaxisDataAccessor.get(newCell->extraAttribPtr);
+			mitr = chemotaxisDataDictRef.find(fieldNameVec[i]);
+			
+			ChemotaxisData * chemotaxisDataPtr = 0;
+			if (mitr != chemotaxisDataDictRef.end()) chemotaxisDataPtr=&mitr->second;
+			
+			// when chemotaxis is allowed towards this type of oldCell and lambda is non-zero
+			if(chemotaxisDataPtr && chemotaxisDataPtr->okToChemotact(oldCell, newCell)){ 
+				ChemotaxisPlugin::chemotaxisEnergyFormulaFcnPtr_t formulaCurrentPtr = 0;
+				formulaCurrentPtr=chemotaxisDataPtr->formulaPtr;
+				if(formulaCurrentPtr){
+					energy += (this->*formulaCurrentPtr)(fieldVec[i]->get(potts->getFlipNeighbor()), fieldVec[i]->get(pt), *chemotaxisDataPtr);
+					chemotaxisDone = true;
+				}
+			}
+
+			if(!chemotaxisDone && (int)newCell->type < vecVecChemotaxisData[i].size()){
+				ChemotaxisData & chemotaxisDataRef = vecVecChemotaxisData[i][(int)newCell->type];
+				ChemotaxisData * chemotaxisDataPtr = & vecVecChemotaxisData[i][(int)newCell->type];
+				ChemotaxisPlugin::chemotaxisEnergyFormulaFcnPtr_t formulaCurrentPtr = 0;
+
+				formulaCurrentPtr = chemotaxisDataRef.formulaPtr;
+
+				if(chemotaxisDataRef.okToChemotact(oldCell, newCell) && chemotaxisDataRef.lambda!=0.0 && formulaCurrentPtr)
+					energy += (this->*formulaCurrentPtr)(fieldVec[i]->get(potts->getFlipNeighbor()), fieldVec[i]->get(pt), chemotaxisDataRef);
+			}
+		}
+		if(oldCell){
+			bool chemotaxisDone=false;
+
+			// first try "locally defined" chemotaxis
+			std::map<std::string,ChemotaxisData> & chemotaxisDataDictRef = *chemotaxisDataAccessor.get(oldCell->extraAttribPtr);
+			mitr = chemotaxisDataDictRef.find(fieldNameVec[i]);
+			
+			ChemotaxisData * chemotaxisDataPtr = 0;
+			if (mitr != chemotaxisDataDictRef.end()) chemotaxisDataPtr=&mitr->second;
+			
+			// when chemotaxis is allowed towards this type of newCell and lambda is non-zero
+			if(chemotaxisDataPtr && chemotaxisDataPtr->okToChemotact(newCell, oldCell)){ 
+				ChemotaxisPlugin::chemotaxisEnergyFormulaFcnPtr_t formulaCurrentPtr = 0;
+				formulaCurrentPtr = chemotaxisDataPtr->formulaPtr;
+				if(formulaCurrentPtr){
+					energy += (this->*formulaCurrentPtr)(fieldVec[i]->get(potts->getFlipNeighbor()), fieldVec[i]->get(pt), *chemotaxisDataPtr);
+					chemotaxisDone = true;
+				}
+			}
+
+			if(!chemotaxisDone && (int)oldCell->type < vecVecChemotaxisData[i].size()){

```

Comment:
I would try to encapsulate lines 621-628 and 590-597 as a separate function and call it instead of repeating the code. In general I would introduce more functions like that. And as you can see that would apply to many CC3D plugins which means we do need need to do it now but slowly work towards eliminating code duplication from many CC3D modules